### DISCUSSION TITLE
Is it possible to use persisted data from localStorage?

### DISCUSSION THREAD

**Discussion by jvnm-dev** [remix-run/remix#1202] [Q&A] [answered]
I have a Remix app using a [zustand](https://github.com/pmndrs/zustand) store. This store is persisted in browser's localStorage.
The use-case is to persist a boolean named `darkModeEnabled`.

### My issue
My code looks like this:
```typescript
const MyComponent = () => {
   const { darkModeEnabled } = useStore(...)

   console.log(darkModeEnabled)

   return darkModeEnabled ? <Moon /> : <Sun />
}
```

After a refresh, `MyComponent` is rendered as if the value in the store were `false` so it shows `<Sun />` but the console.log shows `true` in the console.

I read that this kind of issue is linked to SSR but I don't know how to adapt my code to support my persisted data.

Does someone have an idea perhaps?

**Comment by hollandThomas** [comment, marked as answer]
When server-side-rendering, your code is executed in a JS environment that is not a browser (typically Node.js or Deno or Cloudflare Workers or ...). In particular, those environments do not have access to browser-only APIs such as `localStorage`.

What you need to do is to wrap your calls to browser-only APIs in `useEffect`s. Remix has an example in their docs
https://remix.run/docs/en/v1/guides/constraints#rendering-with-browser-only-apis

**Reply by hollandThomas** [reply]
Though you might want to take an entirely different route (no pun intended) and ditch zustand altogether by splitting your CSS files with the `prefers-color-scheme` media feature.
https://remix.run/docs/en/v1/guides/styling#link-media-queries

Then, use something like this `<dark-mode-toggle>` by GoogleChromeLabs which explicitly recommends the same pattern
https://github.com/GoogleChromeLabs/dark-mode-toggle#-using-different-stylesheets-per-color-scheme-that-are-conditionally-loaded

Note that the toggle is a custom element which I haven't worked with myself. React historically has a couple of problems with those, although that's changing with React 18 from my understanding. In any case, you could give it a try and see if it works as expected
https://github.com/facebook/react/issues/11347#issuecomment-988970952
https://reactjs.org/docs/web-components.html#using-web-components-in-react
https://css-tricks.com/3-approaches-to-integrate-react-with-custom-elements/
